Fix path comparison and include additional guidance when secrets detected

sharon-wang opened the pull request:
<!-- Thank you for submitting a pull request.
If this is your first pull request you can find information about
contributing here:
  * https://github.com/posit-dev/positron/blob/main/CONTRIBUTING.md

We recommend synchronizing your branch with the latest changes in the
main branch by either pulling or rebasing.
-->

<!--
  Describe briefly what problem this pull request resolves, or what
  new feature it introduces. Include screenshots of any new or altered
  UI. Link to any GitHub issues but avoid "magic" keywords that will 
  automatically close the issue. If there are any details about your 
  approach that are unintuitive or you want to draw attention to, please 
  describe them here.
-->

- Addresses: https://github.com/posit-dev/positron/issues/3413#issuecomment-2195084061

### Fixes in this PR

- Fix path comparison issue on Windows
- Note in console output that `detect-secrets` should be installed if it isn't already
    <img width="887" alt="image" src="https://github.com/posit-dev/positron/assets/25834218/76241822-ed7e-4d75-a8ed-902198133175">

- Add some guidance on what actions to take when secrets are detected
    <img width="879" alt="image" src="https://github.com/posit-dev/positron/assets/25834218/8e80c8e9-199d-40fd-96e9-ce376718fd3c">

### QA Notes

<!--
  Add additional information for QA on how to validate the change,
  paying special attention to the level of risk, adjacent areas that
  could be affected by the change, and any important contextual
  information not present in the linked issues.
-->

See https://github.com/posit-dev/positron/pull/3618 for more detail on testing and some example secrets to try.

#### Suggested tests
- test the hook on Windows
- test the hook without detect-secrets installed
- some new output should show when secrets are detected, which guide the user to remove the secrets / mark them as false positives / consult the secrets README

sharon-wang committed:
compare resolved paths for better platform independency

On Windows, we can have the same path but with different directory separators, so if we do a string comparison, the paths are considered to be different. Instead, we should compare the resolved paths which are normalized and have the same directory separator.

sharon-wang committed:
throw detect-secrets errors so error messages are propagated

sharon-wang committed:
add guidance on what to do if secrets are found

Outline actions to take in console output when secrets are found

sharon-wang committed:
make console output spacing a bit nicer

sharon-wang committed:
add some more console output messages and fix up formatting

sharon-wang committed:
add some comments to explain stdio param and no-verify flag

sharon-wang commented on build/detect-secrets.js:
```diff
@@ -41,10 +41,11 @@ const execSync = (command, stdio = ['inherit', 'pipe', 'inherit']) => {
-	const gitRepoRoot = execSync(gitCommand);
+	const gitRepoRoot = execSync(gitCommand).trim();
 	printDebug(gitRepoRoot);
-	printDebug('current working directory: ' + process.cwd());
-	if (process.cwd().trim() !== gitRepoRoot.trim()) {
+	const currentDir = process.cwd().trim();
+	printDebug('current working directory: ' + currentDir);
+	if (path.resolve(currentDir) !== path.resolve(gitRepoRoot)) {
```
this fixes the Windows path comparison issue

sharon-wang commented on build/detect-secrets.js:
```diff
@@ -107,9 +118,13 @@ const detectSecretsScan = (args, stdio) => {
 // Ensure that detect-secrets is installed
 try {
-	detectSecrets('--version');
+	// pipe the stdio so we print our own error message if detect-secrets is not installed
+	detectSecrets('--version', stdio = 'pipe', throwIfError = true);
 } catch (error) {
+	printDebug(error);
+	console.error(); // print newline
 	console.error(`${'Error:'.red} detect-secrets is not installed. Install detect-secrets with ${'pip install detect-secrets'.magenta} or ${'brew install detect-secrets'.magenta}.`);
+	console.error(); // print newline
 	process.exit(ExitCodes.DETECT_SECRETS_WRAPPER_ERROR);
```
this, along with the changes above to include `throwIfError`, ensure we'll print the note to install `detect-secrets` if applicable when the precommit hook is run

sharon-wang commented on build/detect-secrets.js:
```diff
@@ -164,23 +181,37 @@ const runDetectSecretsHook = () => {
+			if (error.stdout) {
+				console.error(error.stdout.toString().red);
+			}
+			// print guidance on how to manage the possible secrets
+			console.error('Uh oh! If you have secrets in your code, please remove them before committing.'.magenta);
+			console.error(`If you are certain that these are false positives, see ${'build/secrets/README.md'.underline} for instructions on how to mark them as such.`.magenta);
+			console.error(); // print newline
```
added guidance on what to do if secrets are detected

sharon-wang reviewed (commented):
I have some file formatting fixes to apply as well, but I'll do that after this has been reviewed.

nstrayer commented on build/detect-secrets.js:
```diff
@@ -107,9 +118,13 @@ const detectSecretsScan = (args, stdio) => {
+	detectSecrets('--version', stdio = 'pipe', throwIfError = true);
 } catch (error) {
+	printDebug(error);
+	console.error(); // print newline
 	console.error(`${'Error:'.red} detect-secrets is not installed. Install detect-secrets with ${'pip install detect-secrets'.magenta} or ${'brew install detect-secrets'.magenta}.`);
+	console.error(); // print newline
```
Not terribly important but you can avoid these extra calls. 

```suggestion
	console.error(`\n${'Error:'.red} detect-secrets is not installed. Install detect-secrets with ${'pip install detect-secrets'.magenta} or ${'brew install detect-secrets'.magenta}.\n`);
```

nstrayer commented on build/detect-secrets.js:
```diff
@@ -130,8 +145,10 @@ const baselineFileExists = () => {
 	if (!baselineFileExists()) {
-		console.error(`${'Error:'.red} Baseline file ${baselineFile.underline} does not exist.
-Run ${'node build/detect-secrets.js init-baseline'.magenta} to create it.`);
+		console.error(); // print newline
+		console.error(`Error: Baseline file ${baselineFile.underline} does not exist.`.red);
+		console.error(`Run ${'node build/detect-secrets.js init-baseline'.magenta} to create it.`);
+		console.error(); // print newline
```
Another example. 
```suggestion
		console.error(
			`\nError: Baseline file ${baselineFile.underline} does not exist.\n`.red +
			`Run ${'node build/detect-secrets.js init-baseline'.magenta} to create it.\n`
		);
```

nstrayer commented on build/detect-secrets.js:
```diff
@@ -17,10 +17,10 @@ const colors = require('colors');
 	FOUND_SECRETS_OR_BASELINE_ISSUE: 1, // detect-secrets-hook found secrets or encountered a baseline file issue
 	DETECT_SECRETS_WRAPPER_ERROR: 2,    // an error occurred in this script
-	DETECT_SECRETS_ERROR: 3,            // an error occurred in detect-secrets
+	DETECT_SECRETS_ERROR: 999,          // an error occurred in detect-secrets (detect-secrets status code is appended)
```
What's the reason for updating this to `999`? Is `999+n` a special status code for the hygiene checks?

sharon-wang replied:
Previously, I was using `DETECT_SECRETS_ERROR: 3` as a prefix for the actual status code returned by `detect-secrets`, but I was using the format `<prefix>_<detect-secrets-code>` (e.g. `3_1`) which is not an integer and thus invalid to return.

So `999` is just to make it more visually clear in the format `<prefix><detect-secrets-code>`, e.g. `9991`

sharon-wang replied:
`detect-secrets` also uses the status codes 0 and 1 and least, and I wanted to differentiate between this wrapper script returning status code 1 and `detect-secrets` returning status code 1 (which would be returned as 9991 by the wrapper)

nstrayer replied:
Can exit codes be that high? I get that it's a string here but eventually, it becomes an 8-bit integer, right?

sharon-wang replied:
Haha yeah maybe they can't be that high :) It seems like the limit for a status code might be 255? 

I just set it back to `3` and removed the appending of the `detect-secrets` status code. The console error output and debug prints will have that extra information if needed.

Thanks for checking this!!

nstrayer reviewed (approved):
Looks good to me. Just one question about the detect-secrets status code and a note about potentially reducing the number of calls to `console.error()` (I'm 50-50 though on if they actually make the code better.)

sharon-wang committed:
consolidate consecutive `console` output calls

although minor in this context, there is a performance penalty when calling `console` log/error/etc more times than necessary.

sharon-wang committed:
apply prettier formatting to `detect-secrets.js`

sharon-wang committed:
don't append `detect-secrets` status code

our exit codes may end up being too high to be valid return status integers if we combine them with `detect-secrets` status codes.
